### PR TITLE
Update: Allow parser to be relative to config (fixes #4985)

### LINKED ISSUE #4985: Cannot find parsers/plugins if not in parent module's node_modules
Some background: I am creating a build system that utilizes plugins. These plugins reside in separate directories from the main module. One of these plugins lints files using ESLint. It needs to support custom parsers and plugins.

I have a system where a **module A** (which is an executable, run at **directory A**) requires a **module B** which then requires ESLint. Thus, ESLint's working directory is **directory A**, so when it goes to look for a `.eslintrc`, it looks in **directory A**, not in **module B**'s directory (this is okay). However, when it attempts to require() parsers or plugins specified in that same `.eslintrc`, it looks in **module B**'s directory. It fails with an error because these parsers/plugins are not present in **module B**'s directory, only in **directory A** (which is also where the `.eslintrc` resides). Not only is this inconsistent behavior (looking for configuration files and parsers/plugins in different directories), it also makes it impossible for me to support arbitrary parsers/plugins in my tool. I do not know if this problem can be fixed because the inconsistent behavior [seems to be the intended behavior](https://github.com/nodejs/node-v0.x-archive/issues/1822)).

The code in ESLint causing this behavior seems to be [this line](https://github.com/eslint/eslint/blob/1e1b456ce84e0058a607c326fe9de3be68f8feab/lib/eslint.js#L528). I am using eslint 1.10.3, and I am using it through the Node.js API, not through the command line.

Is there any way to fix this or is my use case just too complex?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29959747-cannot-find-parsers-plugins-if-not-in-parent-module-s-node_modules?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by nzakas**

### CODE REVIEW

File: docs/developer-guide/working-with-rules.md [hunk 1]
```diff
@@ -74,7 +74,7 @@ The `context` object contains additional functionality that is helpful for rules
 * `id` - the rule ID.
 * `options` - an array of rule options.
 * `settings` - the `settings` from configuration.
-* `parserName` - the `parser` from configuration.
+* `parserPath` - the full path to the `parser` from configuration.
```
Comment (nzakas) [inline docs/developer-guide/working-with-rules.md]: Changed from `parserName` to better reflect the value that's now there. (`parserName` was added in previous beta.)

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,4 @@
-node_modules/
+/node_modules
```
Comment (nzakas) [inline .gitignore]: Had to do this to allow a `node_modules` directory in fixtures.

File: lib/config/config-file.js [hunk 3]
```diff
@@ -395,6 +396,14 @@ function load(filePath, applyEnvironments) {
             Plugins.loadAll(config.plugins);
         }
 
+        // include full path of parser if present
+        if (config.parser) {
+            config.parser = resolveModule.sync(config.parser, {
+                basedir: path.dirname(filePath),
```
Comment (mysticatea) [inline lib/config/config-file.js]: Though I'm not sure it's a problem here, but I just remembered https://github.com/substack/node-resolve/issues/78

Comment (nzakas) [inline lib/config/config-file.js]: Ooh, good catch. I'll wrap that with path.resolve

**Comment by nzakas** [comment]
Crap. somehow managed to combine two different PRS into this one. :( My git fu is weak, so I'm not sure how to separate them back out again.

**Comment by gyandeeps** [comment]
you can create a branch by sha for the commits

**Comment by nzakas** [comment]
Oooh, I'll try that.

**Comment by ilyavolodin** [comment]
Or you will have to cherry-pick changes into a new branch.

**Comment by nzakas** [comment]
It worked. As usual, @gyandeeps saves the day. :)

**Comment by gyandeeps** [comment]
:+1:

**Pull request merged**